realsense/plugin.cpp

```diff
@@ -0,0 +1,179 @@
+#include <librealsense2/rs.hpp> // Include RealSense Cross Platform API
+#include <opencv2/opencv.hpp>   // Include OpenCV API
+#include <mutex>
+
+// ILLIXR includes
+#include "common/threadloop.hpp"
+#include "common/switchboard.hpp"
+#include "common/data_format.hpp"
+
+using namespace ILLIXR;
+
+static constexpr int IMAGE_WIDTH = 640;
+static constexpr int IMAGE_HEIGHT = 480;
+static constexpr int FPS = 30;
+static constexpr int GYRO_RATE = 400; // 200 or 400
+static constexpr int ACCEL_RATE = 250; // 63 or 250
+
+typedef struct {
+	cv::Mat* img0;
+	cv::Mat* img1;
+	cv::Mat* rgb;
+	cv::Mat* depth;
+	int iteration;
+} cam_type;
+
+typedef struct {
+	rs2_vector* accel_data;
+	int iteration;
+} accel_type;
+
+class realsense : public plugin {
+public:
+	realsense(std::string name_, phonebook *pb_)
+	: plugin{name_, pb_}
+	, sb{pb->lookup_impl<switchboard>()}
+	, _m_imu_cam{sb->publish<imu_cam_type>("imu_cam")}
+	, _m_rgb_depth{sb->publish<rgb_depth_type>("rgb_depth")}
+	{
+		auto callback = [&](const rs2::frame& frame)
+    {
+				std::lock_guard<std::mutex> lock(mutex);
+
+				if (auto fs = frame.as<rs2::frameset>()) {
+					rs2::video_frame ir_frame_left = fs.get_infrared_frame(1);
+					rs2::video_frame ir_frame_right = fs.get_infrared_frame(2);
+					rs2::video_frame depth_frame = fs.get_depth_frame();
+					rs2::video_frame rgb_frame = fs.get_color_frame();
+					cv::Mat ir_left = cv::Mat(cv::Size(IMAGE_WIDTH, IMAGE_HEIGHT), CV_8UC1, (void*)ir_frame_left.get_data());
+					cv::Mat ir_right = cv::Mat(cv::Size(IMAGE_WIDTH, IMAGE_HEIGHT), CV_8UC1, (void*)ir_frame_right.get_data());
+					cv::Mat rgb = cv::Mat(cv::Size(IMAGE_WIDTH, IMAGE_HEIGHT), CV_8UC3, (void*)rgb_frame.get_data());
+					cv::Mat depth = cv::Mat(cv::Size(IMAGE_WIDTH, IMAGE_HEIGHT), CV_16UC1, (void*)depth_frame.get_data());
+					cv::Mat converted_depth;
+					depth.convertTo(converted_depth, CV_32FC1, depth_scale * 1000.f);
+
+					cam_type_.img0 = new cv::Mat{ir_left};
```

Comment:
Declaring these all at once is slightly preferred:

```C++
cam_type_ = cam_type{
            // Make a copy, so that we don't have race
            new cv::Mat{imageL_ocv},
            new cv::Mat{imageR_ocv},
            new cv::Mat{rgb_ocv},
            new cv::Mat{depth_ocv},
            iteration_no,
        };
```